src/main/include/utils/StateManager.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+
+#include <frc2/command/CommandPtr.h>
+#include <frc2/command/CommandScheduler.h>
+#include <frc2/command/DeferredCommand.h>
+#include <frc2/command/InstantCommand.h>
+#include <frc2/command/SubsystemBase.h>
+
+template <typename StateType>
+class StateManager : public frc2::SubsystemBase {
+ public:
+  StateManager(const std::function<bool()> Interrupt, StateType DefaultState,
+               frc2::CommandPtr&& DefaultEntrypoint)
+      : m_currentState(DefaultState),
+        m_defaultState(DefaultState),
+        m_defaultEntryPoint(DefaultEntrypoint){};
+
+  /**
+   * Register a new state
+   *
+   * @param StateType A type that represents the state
+   * @param Entrypoint A CommandPtr that will run while the state is active
+   * @param Timeout A timeout for the Entrypoint, after it has elapsed the state
+   * will be canceled
+   */
+  void RegisterState(StateType State, frc2::CommandPtr&& Entrypoint,
+                     units::second_t Timeout, bool ReturnToDefault = false) {
+    auto cmd = std::move(Entrypoint).WithTimeout(Timeout);
+    if (ReturnToDefault) {
+      cmd = cmd.AndThen(SetState(m_defaultState));
+    }
+    m_cmds[State] = cmd;
+  };
+
+  /**
+   * Register a new state
+   *
+   * @param StateType A type that represents the state
+   * @param Entrypoint A CommandPtr that will run while the state is active
+   */
+  void RegisterState(StateType State, frc2::CommandPtr&& Entrypoint,
+                     bool ReturnToDefault = false) {
+    auto cmd = std::move(Entrypoint);
+    if (ReturnToDefault) {
+      cmd = cmd.AndThen(SetState(m_defaultState));
+    }
+    m_cmds[State] = cmd;
+  };
+
+  /**
+   * Register a new state
+   *
+   * @param StateType A type that represents the state
+   * @param Entrypoint A std::function that will run while the state is active
+   */
+  void RegisterState(StateType State, std::function<void()> Entrypoint,
```

Comment:
Since this is now a function instead of a `CommandPtr`, maybe call it `action` instead? Or match what frc calls it in their `InstantCommand`